### DISCUSSION TITLE
Does APIRoute read on the order level of the router that being added?

### DISCUSSION THREAD

**Discussion by yyoel** [fastapi/fastapi#7284] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
_main.py_

import uvicorn
from app.core.Auth import get_current_user
from fastapi import FastAPI, Depends
from app.settings import API_VERSION
from app.routers import AuthRouter, DTTOTRouter, UserRouter


app = FastAPI()


prefix = "/api/{}".format(API_VERSION)

app.include_router(AuthRouter.router,
                   prefix=prefix,
                   tags=["Auth"])
app.include_router(UserRouter.router,
                   prefix=prefix+'/user',
                   tags=["User"])
app.include_router(DTTOTRouter.router,
                   prefix=prefix+'/dttot',
                   dependencies=[Depends(get_current_user)],
                   tags=["DTTOT"])

if __name__ == "__main__":
    uvicorn.run('main:app', reload=True)
```

```Python
from fastapi import Depends, APIRouter, Header
from typing import Optional

from app.databases.Connection import get_db, Session
from app.core import DTTOT as dttotCore

router = APIRouter()


@router.get('/')
def get_all_dttot(db: Session = Depends(get_db), name: Optional[str] = None):
    return dttotCore.get_all_dttot(db=db, name=name)


@router.get('/{id}')
def get_dttot_by_id(id: int, db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_id(db=db, id=id)


@router.get('/jaro-winkler')
def get_dttot_by_jaro_winkler(name: str = Header(...), db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_jaro_winkler(db=db, name=name)

```

### Description

Hello, first of all, I'm sorry if the title were too confusing, I really don't know if there were bugs or maybe my implementation of APIRoute is just wrong in all way,

I attach some of the code when I use APIRoute in the Example above,

when I order my code like the above example and try to request the API(_the third route_), using curl terminal or from swagger it returns this response

```

422 | Error: Unprocessable Entity
{
  "detail": [
    {
      "loc": [
        "path",
        "id"
      ],
      "msg": "value is not a valid integer",
      "type": "type_error.integer"
    }
  ]
}
``` 
if using curl,
```
curl -X GET "http://localhost:8000/api/v1/dttot/jaro-winkler" -H  "accept: application/json" -H  "name: a" -H  "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ5dWxpdXMiLCJleHAiOjE1OTk1NDkxODZ9.-iPXwmWGEMlW6gTZh8qWWQ21SmFmlTJQwf32c97eTO0"

{"detail":[{"loc":["path","id"],"msg":"value is not a valid integer","type":"type_error.integer"}]}

```
Okay, I don't know where and why the value that I gave becomes 'int' type, I literally giving it an 'str' type of value,

But when I try to change the order like this,

```Python
...

@router.get('/jaro-winkler')
def get_dttot_by_jaro_winkler(name: str = Header(...), db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_jaro_winkler(db=db, name=name)

@router.get('/{id}')
def get_dttot_by_id(id: int, db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_id(db=db, id=id)
```
Suddenly the returns response the right one(_200 OK_)

So back to the question in the title, that could it be APIRoute really read the order of the route, or is it just my implementations that were wrong? 

Regards,
### Environment

* OS:  Windows:
* FastAPI Version 0.61.1:

**Comment by luebke-dev** [comment, marked as answer]
The order is important.
```
@router.get('/jaro-winkler')
def get_dttot_by_jaro_winkler(name: str = Header(...), db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_jaro_winkler(db=db, name=name)

@router.get('/{id}')
def get_dttot_by_id(id: int, db: Session = Depends(get_db)):
    return dttotCore.get_dttot_by_id(db=db, id=id)
```
is the correct way to do that.

if you have a route with a fixed path, you need to define it before the route with a dynamic path.

check out the documentation:
https://fastapi.tiangolo.com/tutorial/path-params/#order-matters

**Comment by yyoel** [comment]
> The order is important.
> 
> ...
> check out the documentation:
> https://fastapi.tiangolo.com/tutorial/path-params/#order-matters

gosh, I skip that one most important doc, thank you for the reminder and the respons @SebastianLuebke !

**Comment by yyoel** [comment]
check out the documentation:
https://fastapi.tiangolo.com/tutorial/path-params/#order-matters

**Comment by tiangolo** [comment]
Thanks for the help here @SebastianLuebke ! :clap: :bow: 

Thanks for reporting back and closing the issue @yyoel  :+1:
